File: src/main/java/com/flowcommerce/core/domain/member/service/MemberService.java
package com.flowcommerce.core.domain.member.service;

import com.flowcommerce.core.domain.member.MemberPolicyValidator;
import com.flowcommerce.core.domain.member.SignUp;
import com.flowcommerce.core.storage.member.MemberEntity;
import com.flowcommerce.core.storage.member.MemberRepository;
import com.flowcommerce.core.support.auth.PasswordEncoder;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class MemberService {
    private final MemberPolicyValidator memberPolicyValidator;
    private final PasswordEncoder passwordEncoder;
    private final MemberRepository memberRepository;

    @Transactional
    public Long signUp(SignUp signUp) {
        memberPolicyValidator.validateNew(signUp);

        return memberRepository.save(
                MemberEntity.forLocalSignUp(
                        signUp.email(),
                        passwordEncoder.encode(signUp.password()),
                        signUp.name()
                )
        ).getId();
    }
}

File: src/main/kotlin/com/flowcommerce/core/domain/member/service/MemberService.kt
package com.flowcommerce.core.domain.member.service

import com.flowcommerce.core.domain.member.MemberPolicyValidator
import com.flowcommerce.core.domain.member.SignUp
import com.flowcommerce.core.storage.member.MemberEntity
import com.flowcommerce.core.storage.member.MemberRepository
import com.flowcommerce.core.support.auth.PasswordEncoder
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class MemberService(
    private val memberPolicyValidator: MemberPolicyValidator,
    private val passwordEncoder: PasswordEncoder,
    private val memberRepository: MemberRepository
) {
    @Transactional
    fun signUp(signUp: SignUp): Long {
        memberPolicyValidator.validateNew(signUp)

        return memberRepository.save(
            MemberEntity.forLocalSignUp(
                signUp.email,
                passwordEncoder.encode(signUp.password),
                signUp.name
            )
        ).id
    }
}
